retrieval/scrape.go

```diff
@@ -568,26 +654,63 @@ loop:
 			switch err {
 			case nil:
 			case errSeriesDropped:
+				err = nil
+				continue
+			case storage.ErrOutOfOrderSample:
+				err = nil
+				log.With("timeseries", string(met)).Warn("Out of order sample")
+				continue
+			case storage.ErrDuplicateSampleForTimestamp:
+				err = nil
+				log.With("timeseries", string(met)).Warn("Duplicate sample for timestamp")
 				continue
 			default:
 				break loop
 			}
 			// Allocate a real string.
 			mets = string(met)
-			sl.cache[mets] = ref
+			sl.refCache[mets] = ref
+			str := lset.String()
+			sl.lsetCache[ref] = lsetCacheEntry{lset: lset, str: str}
+			if tp == nil {
+				// Bypass staleness logic if there is an explicit timestamp.
+				samplesScraped[str] = lset
+			}
 		}
 		added++
 	}
 	if err == nil {
 		err = p.Err()
 	}
+	if err == nil {
+		for metric, lset := range sl.samplesInPreviousScrape {
+			if _, ok := samplesScraped[metric]; !ok {
+				// Sample no longer exposed, mark it stale.
+				_, err = app.Add(lset, defTime, math.Float64frombits(value.StaleNaN))
+				switch err {
+				case nil:
+				case errSeriesDropped:
+					err = nil
+					continue
+				case storage.ErrOutOfOrderSample, storage.ErrDuplicateSampleForTimestamp:
+					// Do not log here, as this is expected if a target goes away and comes back
+					// again with a new scrape loop.
+					err = nil
+					continue
+				default:
+					break
+				}
+			}
+		}
+	}
```

Comment:
If the storage fails for the stale markers and returns an error we should rollback, so I think this is right.

Code is using gofmt.